race-buffer/src/reader.rs

```diff
@@ -0,0 +1,139 @@
+    fn snap_wcurs(&self) -> Result<usize>;
+    /// Take a snapshot of the overwrite cursor of the RaceBuffer
+    fn snap_owcurs(&self) -> Result<usize>;
+    /// Take a snapshot of the RaceBuffer's backing storage at the given index
+    fn snap_storage(&self, index: usize) -> Result<E>;
+}
+
+/// Struct used to read from a RaceBuffer asynchronously
+pub struct RaceBufferReader<E, S>
+where
+    E: Entry,
+    S: Snapper<E>,
+{
+    /// Struct for reading writer state
+    snapper: S,
+    /// Reader's cursor
+    rcurs: usize,
+    /// Capacity of backing storage
+    storage_cap: usize,
+    /// Cached prefix, not put into buffer until suffix is successfully read
+    stored_prefix: Option<E>,
+    phantom: PhantomData<E>,
+}
+
+impl<E, S> RaceBufferReader<E, S>
+where
+    E: Copy + PartialEq + Entry,
+    S: Snapper<E>,
+{
+    /// Create new RaceBufferReader
+    pub fn new(snapper: S, storage_cap: usize) -> RaceBufferReader<E, S> {
+        RaceBufferReader {
+            snapper,
+            rcurs: 0,
+            storage_cap,
+            stored_prefix: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Attempt to read all new entries in race buffer into given read buffer
+    pub fn read(&mut self, rbuf: &mut Vec<Option<E>>) -> Result<()> {
+        let pre_wcurs = self.snapper.snap_wcurs()?;
+        let pre_owcurs = self.snapper.snap_owcurs()?;
+
+        if pre_owcurs > self.rcurs + self.storage_cap {
+            // If writer has overwritten unread entries between reads, store nils and correct rcurs
+            let num_missed = pre_owcurs - (self.rcurs + self.storage_cap);
+            self.store_missed_items(num_missed, rbuf);
+            self.rcurs = pre_owcurs - self.storage_cap;
+        }
+
+        // Perform reads into snapshot buffer up to write cursor
+        let first_read = self.rcurs;
+        let mut buf_snapshot: Vec<E> = Vec::new();
+        while self.rcurs < pre_wcurs {
+            let storage_index = get_cursor_index(self.storage_cap, self.rcurs, false);
+            buf_snapshot.push(self.snapper.snap_storage(storage_index)?);
+            self.rcurs += 1;
+        }
+
+        // Calculate number of entries in snapshot buffer that may have been overwritten
+        let post_owcurs = self.snapper.snap_owcurs()?;
+        let overlap = min(
+            if post_owcurs > self.storage_cap + first_read {
+                post_owcurs - (first_read + self.storage_cap)
+            } else {
+                0
+            },
+            buf_snapshot.len(),
+        );
+
+        if overlap != 0 {
+            self.store_missed_items(overlap, rbuf);
+        }
+
+        // Store valid entries in read buffer
+        for entry in &mut buf_snapshot[overlap..] {
+            self.store(*entry, rbuf);
+        }
+        Ok(())
+    }
+
+    /// Store given entry in given read buffer
+    #[inline]
+    fn store(&mut self, entry: E, rbuf: &mut Vec<Option<E>>) {
+        match self.stored_prefix {
```

Comment:
Unessential tip: Consider using `Option::take` if you have a pattern of trying to use a stored optional value that you intend to replace or clear-out as a result of the operations. E.G.

```rust
if let Some(prefix) = self.stored_prefix.take() {
    rbuf.push(Some(prefix));
    rbuf.push(Some(entry));
} else if entry.is_prefix() {
    self.stored_prefix = Some(entry);
} else {
    rbuf.push(Some(entry));
}
```